File: bindings/ocaml/examples/basic_usage.ml
(* Basic usage example for Gelfgren OCaml bindings *)

open Gelfgren

let () =
  Printf.printf "Gelfgren OCaml Bindings Example\n";
  Printf.printf "%s\n\n" (String.make 60 '=');

  (* Example 1: Bernstein Polynomial *)
  Printf.printf "1. Bernstein Polynomial\n";
  Printf.printf "%s\n" (String.make 60 '-');

  (* Using with_poly for automatic cleanup *)
  BernsteinPolynomial.with_poly [|1.0; 2.0; 6.0|] 0.0 1.0 (fun poly ->
      (* Get degree *)
      let deg = BernsteinPolynomial.degree poly in
      Printf.printf "Created polynomial with degree: %d\n" deg;

      (* Evaluate at a point *)
      let y = BernsteinPolynomial.evaluate poly 0.5 in
      Printf.printf "f(0.5) = %.6f\n" y;

      (* Vectorized evaluation *)
      Printf.printf "\nVectorized evaluation:\n";
      let xs = [|0.0; 0.25; 0.5; 0.75; 1.0|] in
      let ys = BernsteinPolynomial.evaluate_array poly xs in
      Array.iter2
        (fun x y -> Printf.printf "  f(%.2f) = %.6f\n" x y)
        xs ys;

      (* Derivative *)
      Printf.printf "\n";
      let dpoly = BernsteinPolynomial.derivative poly in
      let dy = BernsteinPolynomial.evaluate dpoly 0.5 in
      Printf.printf "Derivative at x=0.5: %.6f\n" dy;
      BernsteinPolynomial.free dpoly;

      (* Integral *)
      let ipoly = BernsteinPolynomial.integral poly in
      let y1 = BernsteinPolynomial.evaluate ipoly 1.0 in
      let y0 = BernsteinPolynomial.evaluate ipoly 0.0 in
      let area = y1 -. y0 in
      Printf.printf "Integral from 0 to 1: %.6f\n" area;
      BernsteinPolynomial.free ipoly
    );

  Printf.printf "\n";

  (* Example 2: Polynomial Arithmetic *)
  Printf.printf "2. Polynomial Arithmetic\n";
  Printf.printf "%s\n" (String.make 60 '-');

  let p1 = BernsteinPolynomial.create [|1.0; 2.0; 3.0|] 0.0 1.0 in
  let p2 = BernsteinPolynomial.create [|2.0; 1.0; 1.0|] 0.0 1.0 in

  let y1 = BernsteinPolynomial.evaluate p1 0.5 in
  let y2 = BernsteinPolynomial.evaluate p2 0.5 in
  Printf.printf "p1(0.5) = %.6f\n" y1;
  Printf.printf "p2(0.5) = %.6f\n" y2;

  (* Addition using infix operator *)
  let open BernsteinPolynomial in
  let sum = p1 + p2 in
  let y_sum = evaluate sum 0.5 in
  Printf.printf "p1 + p2 at x=0.5: %.6f\n" y_sum;
  free sum;

  (* Scaling using infix operator *)
  let scaled = 2.0 *. p1 in
  let y_scaled = evaluate scaled 0.5 in
  Printf.printf "2 * p1 at x=0.5: %.6f\n" y_scaled;
  free scaled;

  free p1;
  free p2;

  Printf.printf "\n";

  (* Example 3: Rational Function *)
  Printf.printf "3. Rational Function\n";
  Printf.printf "%s\n" (String.make 60 '-');

  (* Create rational function: x / (1 + x) on [0, 1] *)
  let num = BernsteinPolynomial.create [|0.0; 1.0|] 0.0 1.0 in
  let den = BernsteinPolynomial.create [|1.0; 2.0|] 0.0 1.0 in

  RationalFunction.with_rational num den (fun rat ->
      let x = 0.5 in
      let y = RationalFunction.evaluate rat x in
      let expected = x /. (1.0 +. x) in
      Printf.printf "R(%.1f) = %.6f\n" x y;
      Printf.printf "Expected: %.6f\n" expected;

      (* Vectorized evaluation *)
      Printf.printf "\nVectorized evaluation:\n";
      let xs = [|0.0; 0.25; 0.5; 0.75; 1.0|] in
      let ys = RationalFunction.evaluate_array rat xs in
      Array.iter2 (fun x y ->
          let expected = x /. (1.0 +. x) in
          Printf.printf "  R(%.2f) = %.6f (expected: %.6f)\n" x y expected
        ) xs ys
    );

  BernsteinPolynomial.free num;
  BernsteinPolynomial.free den;

  Printf.printf "\n";

  (* Example 4: Padé Approximant *)
  Printf.printf "4. Padé Approximant\n";
  Printf.printf "%s\n" (String.make 60 '-');

  (* Approximate exp(x) near x=0 with [2/2] Padé approximant *)
  let coeffs = [|1.0; 1.0; 0.5; 1.0/.6.0; 1.0/.24.0|] in

  PadeApproximant.with_pade coeffs 2 2 (-1.0) 1.0 (fun pade ->
      Printf.printf "Approximation of exp(x):\n";
      let xs = [|-0.5; -0.25; 0.0; 0.25; 0.5|] in
      let ys = PadeApproximant.evaluate_array pade xs in
      Array.iter2 (fun x y ->
          let exact = exp x in
          let error = abs_float (y -. exact) in
          Printf.printf "  x=%.2f: Padé=%.6f, exp(x)=%.6f, error=%.8f\n"
            x y exact error
        ) xs ys
    );

  Printf.printf "\n";
  Printf.printf "All examples completed successfully!\n"

File: bindings/julia/examples/basic_usage.jl
#!/usr/bin/env julia

# Basic usage example for Gelfgren Julia bindings

using Gelfgren

println("Gelfgren Julia Bindings Example")
println("=" ^ 60)
println()

# Example 1: Bernstein Polynomial
println("1. Bernstein Polynomial")
println("-" ^ 60)

# Create a quadratic polynomial: 2x^2 + 3x + 1 on [0, 1]
poly = BernsteinPolynomial([1.0, 2.0, 6.0], 0.0, 1.0)

# Get degree
deg = degree(poly)
println("Created polynomial with degree: $deg")

# Evaluate at a point
y = evaluate(poly, 0.5)
println("f(0.5) = $y")

# Vectorized evaluation
println("\nVectorized evaluation:")
xs = [0.0, 0.25, 0.5, 0.75, 1.0]
ys = evaluate(poly, xs)
for (x, y) in zip(xs, ys)
    println("  f($x) = $y")
end

# Derivative
println()
dpoly = derivative(poly)
dy = evaluate(dpoly, 0.5)
println("Derivative at x=0.5: $dy")

# Integral
ipoly = integral(poly)
y1 = evaluate(ipoly, 1.0)
y0 = evaluate(ipoly, 0.0)
area = y1 - y0
println("Integral from 0 to 1: $area")

println()

# Example 2: Polynomial Arithmetic
println("2. Polynomial Arithmetic")
println("-" ^ 60)

p1 = BernsteinPolynomial([1.0, 2.0, 3.0], 0.0, 1.0)
p2 = BernsteinPolynomial([2.0, 1.0, 1.0], 0.0, 1.0)

y1 = evaluate(p1, 0.5)
y2 = evaluate(p2, 0.5)
println("p1(0.5) = $y1")
println("p2(0.5) = $y2")

# Addition
sum_poly = p1 + p2
y_sum = evaluate(sum_poly, 0.5)
println("p1 + p2 at x=0.5: $y_sum")

# Scaling
scaled = 2.0 * p1
y_scaled = evaluate(scaled, 0.5)
println("2 * p1 at x=0.5: $y_scaled")

println()

# Example 3: Rational Function
println("3. Rational Function")
println("-" ^ 60)

# Create rational function: x / (1 + x) on [0, 1]
num = BernsteinPolynomial([0.0, 1.0], 0.0, 1.0)
den = BernsteinPolynomial([1.0, 2.0], 0.0, 1.0)

rat = RationalFunction(num, den)

x = 0.5
y = evaluate(rat, x)
expected = x / (1.0 + x)
println("R($x) = $y")
println("Expected: $expected")

# Vectorized evaluation
println("\nVectorized evaluation:")
xs = [0.0, 0.25, 0.5, 0.75, 1.0]
ys = evaluate(rat, xs)
for (x, y) in zip(xs, ys)
    expected = x / (1.0 + x)
    println("  R($x) = $y (expected: $expected)")
end

println()

# Example 4: Padé Approximant
println("4. Padé Approximant")
println("-" ^ 60)

# Approximate exp(x) near x=0 with [2/2] Padé approximant
# Power series: 1 + x + x^2/2 + x^3/6 + x^4/24
coeffs = [1.0, 1.0, 0.5, 1.0/6.0, 1.0/24.0]
pade = PadeApproximant(coeffs, 2, 2, -1.0, 1.0)

println("Approximation of exp(x):")
xs = [-0.5, -0.25, 0.0, 0.25, 0.5]
ys = evaluate(pade, xs)

for (x, y) in zip(xs, ys)
    exact = exp(x)
    error = abs(y - exact)
    @printf("  x=%.2f: Padé=%.6f, exp(x)=%.6f, error=%.8f\n", x, y, exact, error)
end

println()

# Example 5: Broadcasting and Ranges
println("5. Broadcasting with Ranges")
println("-" ^ 60)

poly = BernsteinPolynomial([1.0, 2.0, 6.0], 0.0, 1.0)

# Evaluate over a range
xs = 0.0:0.1:1.0
ys = evaluate(poly, collect(xs))

println("Evaluated at $(length(ys)) points")
println("  min: $(minimum(ys))")
println("  max: $(maximum(ys))")
println("  mean: $(sum(ys) / length(ys))")

println()

# Example 6: Error Handling
println("6. Error Handling")
println("-" ^ 60)

try
    # This should fail - empty coefficients
    bad_poly = BernsteinPolynomial(Float64[], 0.0, 1.0)
catch e
    if e isa GelfgrenError
        println("Caught expected error: $(e.msg)")
    else
        rethrow(e)
    end
end

println()
println("All examples completed successfully!")
